### ISSUE TITLE
jit.ndi.receive~does not receive NDI HX streams if using GeForce RTX 2070 GPU

### ISSUE THREAD

**Issue by ygreq** [impsnldavid/jit.ndi#23]
Hi folks! Does jit.ndi.receive~ work for you with NDI Bridge? 
I mean I get the NDI streams from the other place using NDI Monitor or OBS, but in Max I just see the streams advertised, but if I select them, I get no image. I tried with 2 different Windows laptops. I am using the latest version of NDI 5 and latest jit.ndi version

Any ideas? What's your experience?
Thank you!

Later Edit: I modified the title to reflect what I found out. See last comment.

**Comment by ygreq** [comment]
So nobody tested this? :/

**Comment by ygreq** [comment]
After numerous tests this is what I found out. 

I have a weird problem where I cannot see NDI HX streams in one application (Cycling 74 Max 8 app), but I can see NDI HX streams in other apps like OBS and NDI Monitor. I can only see black when streaming NDI HX. And I only have this problem on some Windows machines and not others. Through tests and trials I recently found out that this problem is related to the newer Nvidia GeForce drivers. So I believe it is a bug in either the Max 8 object (jit.ndi), in NDI itself (and imported in Max 8) or in the GeForce drivers. 


I don’t have this issue with the laptop that has GeForce GTX 1060 using some older drivers, but I have this issue if I update to the latest drivers. And for the 2 laptops I tested that have a GeForce RTX 2070 GPU,  I didn’t find a driver to fix this issue. 

I am curious to find out your opinion and suggestion on what to do next. And if there any possibilities to fix this issue. 



Here are the premises:
-I have tested NDI on 5 Windows machines. I have the following problem on 2 out the 5 Windows machines. I recently found out that both machines where NDI HX has this issue have an Nvidia GeForce RTX 2070 GPU.
-The problem affects only the Cycling 74 Max application, but not OBS nor NDI Monitor.
-I cannot see the HX steams (even HX streams created on the same machine - for example using Screen Capture HX) in the Max 8 app on those 2 machines. I can see HX streams in the Max 8 app on the other 3 machines. Normal NDI streams just fine.
-On PTZ camera that output HX streams, even though I only see black (so no image), I can control those specific cameras


And here is the longer version:
Here are the tests I did that led me to the fact that this problem is created by newer Nvidia drivers. I recently reinstalled Win 10 and the necessary apps (Max 8, jit.ndi object necessary to run NDI in Max, and NDI Tools) on the 2 laptops again. Windows 10 installed the video drivers by itself. And I recorded the screen just so you can see the weirdness.

I started with the laptop with a Nvidia GeForce GTX 1060 where NDI HX in Max worked before. And after Win 10 finished installing, I immediately installed Max 8, opened Max, then copied jit.ndi. I then installed NDI Tools. After NDI Tools got installed, I opened NDI Monitor and tested to see if I could see the HX streams coming from the 2 PTZ cameras. Everything worked ok as expected. Then I opened Max. It didn't work as I wasn't able to see external NDI sources. I restarted Max and all the sources became available, including HX streams coming from the 2 cameras. I even tested with NDI Screen Capture HX coming from the same laptop. Everything worked as expected. No issues at all. Every software I used was the version. And the GPU drivers were installed automatically by Windows 10. I did not update them. You can see the demo here: https://youtu.be/LUSjnTNOXZ8 . I recommend 2x speed playback.

I then reinstalled Win 10 on the laptop that has an Nvidia GeForce RTX 2070 GPU that had issues previously where Max couldn't see video from NDI HX sources. I followed the same steps and used exactly the same software (including the same Win 10 usb stick): waited for Win 10 to finish installing, installing Max afterwards, copied jit.ndi, installed NDI Tools, opened NDI Monitor, opened Max, closed and reopened Max. And Max again has the same issue. It does not see HX streams from PTZ cameras as well as HX streams coming from NDI Screen Capture HX running on the same laptop. It does see normal NDI streams. And NDI Monitor does see normal as well as HX streams. And again the GPU drivers were installed automatically by Windows 10. I did not update them. You can see a demo here: https://youtu.be/YPMqQW41hDk


After these tests this is what happened. But keep in mind. The laptop where I had no issues with HX streams in Max has a Nvidia GeForce GTX 1060 GPU. The laptop where I had issues with HX streams in Max has a GeForce RTX 2070 GPU. This is important because this is what I found.

Like I was saying earlier, on the laptop with a GeForce GTX 1060, I reinstalled the necessary software for testing HX streaming in Max and everything was working. And like I mentioned, the GPU drivers were installed automatically by Windows. This happened before I had the time to make the tests. So the tests were done with GPU drivers installed automatically by Windows. After the tests I also installed GeForce Experience app and updated the GPU drivers via that app. Device Manager reported I had v30.0.15.1123 at this point. Once I updated the GPU drivers, I realized that Max stopped streaming HX streams. Just like on the other laptop which has a GeForce RTX 2070 GPU. I went to the Device Manager and unrolled the drivers for the GPU card and now the Device Manager reports I have v22.21.13.8205. I restarted the laptop and now I get HX streaming inside of Max once more. So definitely HX streams are affected by the newer GPU drivers.

On the other laptop that boasts a GeForce RTX 2070 GPU card, the drivers installed automatically by Windows are probably newer and this is why they don't work with HX streams in Max 8 either. I will try to update the GPU drivers there as well, but my guess is that I will have to use older drivers in order to make it work. And I hope I can find older drivers.

As an extra note, I recently checked another laptop that had similar issues also has an RTX 2070 GPU card. 

These are the drivers I used and logged after tests:

Laptop 1 Clevo P751 P751DM2-G
Configuration

CPU: Intel® Core™i7 Quad Core Processor i7-6700k (4.0GHz) 8MB Cache 91w
GPU: NVIDIA® GeForce GTX 1060 (N17E-G1) 6GB GDDR5 Video RAM
RAM: DDR4 SO-DIMM 2x16GB HyperX IMPACT 2133MHz SODIMM DDR4

Drivers tested: 
Not OK v30.0.15.1123 (which is the last version for this laptop) 
OK        v22.21.13.8205 (This is the rolled back version after I tested the one above. I believe this is the version that Windows 10 automatically installs)
OK        v21.21.13.7620 from 12/2/2016 (installed from the laptop manufacturer package of drivers)

Laptop 2 Dream Machine
CPU: AMD Ryzen™ 9 3900X 12 Core 3.8GHz Up to 4.6GHz Boost Clock 64MB L3 Cache
GPU: NVIDIA® GeForce® RTX 2070 - 8.0GB GDDR6 Video RAM
RAM: 2x32GB Corsair VENGEANCE 2666MHz SODIMM DDR4

Drivers tested: 
Not OK v27.21.14.5720 from 11/2/2020 (installed automatically by Windows 10 and also available from the laptop manufacturer package of drivers)
Not OK v10.0.19041.868 from 6/21/2006 (rolled back drivers for generic graphics card) 
Not OK v30.0.15.1123 from 1/10.2022
Not OK v27.21.14.6677 from 6/8/2021





Conclusion
For sure the newer GPU drivers are creating this issue where I get black instead of HX streams in Max 8 via jit.ndi.receive~. And I am asking what I should do in this situation other then using older drivers? Should I also talk to Nvidia or maybe Newtek. I have to bring again the fact that in my tests the only app affected by this GPU driver / NDI bug is the jit.ndi object in Max 8. NDI Monitor and OBS work just fine. I need your advice!

Thank you so much!

**Comment by pixsperdavid** [comment]
An important point to note is that there is no guarantee that by testing with NDI tools you are using the same NDI runtime version as jit.ndi uses. jit.ndi works by dynamically loading (at runtime) the NDI runtime DLL pointed at by the NDILIB_REDIST_FOLDER environment variable. Other applications may do this, or they may use a private copy of the NDI runtime contained within their own application folder. It's possible that the NewTek applications don't use a runtime DLL at all and in fact have the runtime statically linked. The NDI HX driver is even more opaque to jit.ndi, it is only interacted with by the NDI runtime library itself. jit.ndi is completely unaware of it's existence or non-existence.

It's helpful that you've managed to track this down to a driver related issue. However I can't currently see any path to solving this in jit.ndi as none of the code interacts with the GPU drivers directly in any way. It seems that the hardware GPU H264 decode implementation in the newer driver version is causing the problem. I'm going to close the issue for now as I can't realistically do anything to resolve this, but if you find any further information which could required code changes to jit.ndi please re-open.

**Comment by ygreq** [comment]
Thank you, David. But I still have some questions. Considering that on a fresh Windows 10 installation I just installed NDI Tools without NDI SDK or NDI Runtime how did jit.ndi managed to work? Where did it find the NDI runtime dlls? Does NDI Tools come with its own runtime?

And I forgot to mention that whenever I stream an HX stream, jit.fpsgui shows that I actually have a 1x1 pixel matrix instead of a fullHD matrix I have with normal NDI streaming. I wanted to mention this in case it gives you or anyone else some pointers.

**Comment by pixsperdavid** [comment]
Installing NDI tools will also install the NDI runtime.

Out of interest, I did a quick examination on NDI Studio Monitor. It does not have any explicit dependency on the NDI runtime DLL, and I also confirmed that it does not load the DLL at application start-up (although it does load the HX driver). I would assume that the rest of the NDI tools are the same, and do not actually use the NDI runtime but instead have the NDI implementation statically linked into their binaries. This means unfortunately that they are not particularly useful for testing as a point of comparison against applications which do use the runtime.
